File: clients/apps/app/utils/price.ts
import { schemas } from '@polar-sh/client'

export const isLegacyRecurringPrice = (
  price: schemas['ProductPrice'] | schemas['LegacyRecurringProductPrice'],
): price is schemas['LegacyRecurringProductPrice'] => 'legacy' in price

export const hasLegacyRecurringPrices = (
  product: schemas['Product'],
): product is schemas['Product'] & {
  prices: schemas['LegacyRecurringProductPrice'][]
} => product.prices.some(isLegacyRecurringPrice)

File: server/polar/checkout/price.py
from collections.abc import Sequence

from polar.models import ProductPrice


def get_default_price(prices: Sequence[ProductPrice]) -> ProductPrice:
    """
    Get the default price for a checkout session, given a list of product prices.

    We should select the static price in priority,
    as it determines the amount and specific behavior, like PWYW
    """
    for price in prices:
        if price.is_static:
            return price
    return prices[0]
